File: test-suite-java/src/test/java/io/micronaut/interceptor/test/identity/LoggingInterceptor.java
package io.micronaut.interceptor.test.identity;

import jakarta.annotation.PostConstruct;
import jakarta.interceptor.AroundConstruct;
import jakarta.interceptor.AroundInvoke;
import jakarta.interceptor.Interceptor;
import jakarta.interceptor.InvocationContext;

@Interceptor
@Logged
public class LoggingInterceptor {

    @AroundConstruct
    public void constructed(InvocationContext context) throws Exception {
        context.proceed();
    }

    @PostConstruct
    public void initialized(InvocationContext context) throws Exception {
        context.proceed();
    }

    @AroundInvoke
    public Object log(InvocationContext context) throws Exception {
        return context.proceed();
    }
}

File: test-suite-kotlin/src/test/kotlin/io/micronaut/interceptor/test/LoggingInterceptor.kt
package io.micronaut.interceptor.test

import jakarta.annotation.Priority
import jakarta.interceptor.AroundInvoke
import jakarta.interceptor.Interceptor
import jakarta.interceptor.InvocationContext

@Interceptor
@Logged
@Priority(Interceptor.Priority.APPLICATION - 10)
open class LoggingInterceptor {

    @AroundInvoke
    open fun log(context: InvocationContext): Any? {
        Calls.recorded += "log in ${context.method.name}"
        try {
            return context.proceed()
        } finally {
            Calls.recorded += "log out"
        }
    }
}
